Prototype for module-internal chain of tasks

makortel opened the pull request:
#### PR description:

This PR adds helpers for a chain of TBB tasks internal to the ExternalWork module. These helpers can be used to implement e.g. a chain of tasks that alternate CPU and GPU work, or a loop of GPU work where the decision of continuing is done in the CPU. This PR also adds `CUDAScopedContextAnalyze` mentioned in https://github.com/cms-patatrack/cmssw/pull/355#issuecomment-504045555.

In short, the `CUDAScopedContextAcquire` can be used to queue a "next task" instead of the `produce()` with
```cpp
void ...::acquire(...) {
  CUDAScopedContextAcquire ctx{...};

  ctx.insertNextTask([this](WaitingTaskWithArenaHolder h) {
    CUDAScopedContextTask ctx{this->ctxState_, std::move(h)};
     ...
   }):
}
``` 
The commit https://github.com/cms-patatrack/cmssw/commit/203cd7d2100b7f942e5af1567ba12addaed7eb6e experiments with the interface by providing directly the `CUDAScopedContextTask` to the user at the cost of interface inconsistency between `acquire()` and the function

```cpp
void ...::acquire(...) {
  CUDAScopedContextAcquire ctx{...};

  ctx.insertNextTask([](CUDAScopedContextTask ctx) {
     ...
   }):
}
``` 
I would like to hear opinions of others which one would you prefer.


#### PR validation:

Added test runs.

makortel commented:
@vkhristenko this is the PR I mentioned

makortel commented:
@VinInn this mechanism can be used also for "re-running a producer with more memory when initial memory was not sufficient"

makortel commented on FWCore/Concurrency/interface/WaitingTaskWithArenaHolder.h:
```diff
@@ -72,5 +72,29 @@ namespace edm {
+            holder.doneWaiting(*excptr);
+            return;
+          }
+          func();
```
I'm a bit unhappy about the replication of `WaitingTaskWithArenaHolder` here, especially since the code-to-be-copied would be only 4 lines.

fwyzard commented:
:construction: Validation running at workergpu11:/mnt/home/fwyzard/validation/run.JeiyqrUwZQ ...

fwyzard commented:
About the `insertNextTask()` syntax, I prefer the second variant, as it is leaner and avoids repeating the same couple of lines.

makortel commented:
Couple of items from a chat with @fwyzard 
- [ ]  Change the order `insertNextTask()` and CUDA async work launches to be more intuitive (the order doesn't really matter)
- [ ] Make the LIFO stack property of the "next task" more clear in the interface (`insertNextTask()` -> `pushNextTask()`) and examples

makortel commented:
@fwyzard I've addressed the points from our discussion (https://github.com/cms-patatrack/cmssw/pull/363#issuecomment-529558507), could you take a look again?

fwyzard commented:
Thanks, the latest set of changes make it clear what the behaviour of the tasks should be.

fwyzard commented:
The only aspect that it's not clear to me is why

> **Note** that the `CUDAService` is **not** available (nor is any other
service) in these intermediate tasks.

(I assume the same holds for other `Service`s as well).

I thought the `Service`s were available to the underlying *thread*, not to the individual *task* ?

makortel commented:
@fwyzard Services are a rather complex subsystem, approximately requiring a valid ServiceToken on the thread calling Services. The framework holds a ServiceToken during the execution of any module methods, but such token does not exist on arbitrary TBB tasks and thus Services can not be used.

Here I decided to be lazy and not support Services in these tasks (and see how far we get with it).
